File: Leetcode/347 - Top K Frequent Elements/solution.cpp
vector<int> topKFrequent(vector<int>& nums, int k) {
    unordered_map<int, int> numsCount;
    for (int& num: nums) {
        numsCount[num]++;
    }

    unordered_map<int, vector<int>> bucket;
    int maxFreq = 0;
    for (auto& [num, freq]: numsCount) {
        maxFreq = max(maxFreq, freq);
        bucket[freq].push_back(num);
    }

    vector<int> res;
    for (int freq = maxFreq; k > 0; freq--) {
        for (int num: bucket[freq]) {
            res.push_back(num);
            k--;
        }
    }

    return res;
}

File: Leetcode/347 - Top K Frequent Elements/solution.js
var topKFrequent = function(nums, k) {
    const numsCount = new Map();
    for (const num of nums) {
        numsCount.set(num, (numsCount.get(num) || 0) + 1);
    }

    const bucket = new Map();
    let maxFreq = 0;
    for (const [num, freq] of numsCount) {
        maxFreq = Math.max(maxFreq, freq);
        if (!bucket.has(freq)) bucket.set(freq, []);
        bucket.get(freq).push(num);
    }

    const res = [];
    for (let freq = maxFreq; k > 0; freq--) {
        if (bucket.has(freq)) {
            for (const num of bucket.get(freq)) {
                res.push(num);
                k--;
            }
        }
    }
    return res;
};